[BUG] fabric_spark_custom_pool: Provider produced inconsistent result after apply

### 🐛 What happened?

When I try to create/update a fabric_spark_custom_pool resource I get the following error:

```text
fabric_spark_custom_pool.example: Destroying... [id=[ID]]
fabric_spark_custom_pool.example: Destruction complete after 3s
fabric_spark_custom_pool.example: Creating...


│ Error: Provider produced inconsistent result after apply
│
│ When applying changes to fabric_spark_custom_pool.example, provider "provider[\"registry.terraform.io/microsoft/fabric\"]" produced an unexpected new value: .dynamic_executor_allocation.max_executors: was cty.NumberIntVal(2), but now   
│ null.
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

I am new to filing issues but the last sentence made me decide to create this issue.

### 🔬 How to reproduce?

1. Create a spark custom pool
2. Update the resource declaration and run 'terraform apply' again



### 🏗️ Code Sample / Log

```terraform
resource "fabric_spark_custom_pool" "example" {
  workspace_id =  fabric_workspace.example_workspace.id
  name         = "small_pool"
  type         = "Workspace"
  node_family  = "MemoryOptimized"
  node_size    = "Small"

  auto_scale = {
    enabled        = false
    max_node_count = 1
    min_node_count = 1
  }

  dynamic_executor_allocation = {
    enabled       = false
    max_executors = 1
    min_executors = 1
  }

  timeouts = {
    read = "10m"
  }
}
```

Here is the (DEBUG) tf.log

```text
provider-fabric_v0.1.0-beta.8.exe: CREATE: @module=fabric action=start tf_rpc=ApplyResourceChange @caller=github.com/microsoft/terraform-provider-fabric/internal/services/spark/resource_spark_custom_pool.go:166 tf_provider_addr=registry.terraform.io/microsoft/fabric tf_req_id=ID tf_resource_type=fabric_spark_custom_pool timestamp="2025-02-06T14:39:08.085+0100"
2025-02-06T14:39:08.087+0100 [INFO]  provider.terraform-provider-fabric_v0.1.0-beta.8.exe: create timeout configuration is null or unknown, using provided default: @module=fabric tf_req_id=ID tf_provider_addr=registry.terraform.io/microsoft/fabric tf_resource_type=fabric_spark_custom_pool tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-framework-timeouts@v0.5.0/resource/timeouts/timeouts.go:145 timestamp="2025-02-06T14:39:08.087+0100"
2025-02-06T14:39:08.678+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: CREATE: @caller=github.com/microsoft/terraform-provider-fabric/internal/services/spark/resource_spark_custom_pool.go:205 @module=fabric tf_provider_addr=registry.terraform.io/microsoft/fabric tf_resource_type=fabric_spark_custom_pool tf_rpc=ApplyResourceChange action=end tf_req_id=ID timestamp="2025-02-06T14:39:08.677+0100"
2025-02-06T14:39:08.678+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwschemadata/value_semantic_equality.go:91 tf_attribute_path=node_family tf_provider_addr=registry.terraform.io/microsoft/fabric tf_req_id=ID tf_resource_type=fabric_spark_custom_pool timestamp="2025-02-06T14:39:08.678+0100"
2025-02-06T14:39:08.678+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: Value switched to prior value due to semantic equality logic: tf_provider_addr=registry.terraform.io/microsoft/fabric tf_req_id=ID tf_resource_type=fabric_spark_custom_pool tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwschemadata/value_semantic_equality.go:91 @module=sdk.framework tf_attribute_path=workspace_id timestamp="2025-02-06T14:39:08.678+0100"
2025-02-06T14:39:08.678+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: Value switched to prior value due to semantic equality logic: tf_req_id=ID tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwschemadata/value_semantic_equality.go:91 tf_provider_addr=registry.terraform.io/microsoft/fabric @module=sdk.framework tf_attribute_path=name tf_resource_type=fabric_spark_custom_pool timestamp="2025-02-06T14:39:08.678+0100"
2025-02-06T14:39:08.678+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_resource_type=fabric_spark_custom_pool tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwschemadata/value_semantic_equality.go:91 tf_attribute_path=node_size tf_provider_addr=registry.terraform.io/microsoft/fabric tf_req_id=ID  timestamp="2025-02-06T14:39:08.678+0100"
2025-02-06T14:39:08.678+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_attribute_path=timeouts.read tf_resource_type=fabric_spark_custom_pool tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwschemadata/value_semantic_equality.go:91 tf_provider_addr=registry.terraform.io/microsoft/fabric tf_req_id=ID timestamp="2025-02-06T14:39:08.678+0100"
2025-02-06T14:39:08.679+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: Value switched to prior value due to semantic equality logic: @caller=github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwschemadata/value_semantic_equality.go:91 @module=sdk.framework tf_attribute_path=timeouts tf_provider_addr=registry.terraform.io/microsoft/fabric tf_req_id=ID tf_resource_type=fabric_spark_custom_pool tf_rpc=ApplyResourceChange timestamp="2025-02-06T14:39:08.678+0100"
2025-02-06T14:39:08.679+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: Value switched to prior value due to semantic equality logic: @caller=github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwschemadata/value_semantic_equality.go:91 @module=sdk.framework tf_attribute_path=auto_scale.max_node_count tf_rpc=ApplyResourceChange tf_provider_addr=registry.terraform.io/microsoft/fabric tf_req_id=ID tf_resource_type=fabric_spark_custom_pool timestamp="2025-02-06T14:39:08.678+0100"
2025-02-06T14:39:08.679+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: Value switched to prior value due to semantic equality logic: tf_attribute_path=auto_scale.enabled tf_resource_type=fabric_spark_custom_pool tf_provider_addr=registry.terraform.io/microsoft/fabric tf_req_id=ID tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwschemadata/value_semantic_equality.go:91 @module=sdk.framework timestamp="2025-02-06T14:39:08.678+0100"
2025-02-06T14:39:08.679+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: Value switched to prior value due to semantic equality logic: @caller=github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwschemadata/value_semantic_equality.go:91 tf_attribute_path=auto_scale.min_node_count tf_req_id=ID @module=sdk.framework tf_provider_addr=registry.terraform.io/microsoft/fabric tf_resource_type=fabric_spark_custom_pool tf_rpc=ApplyResourceChange timestamp="2025-02-06T14:39:08.679+0100"
2025-02-06T14:39:08.679+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_provider_addr=registry.terraform.io/microsoft/fabric tf_req_id=ID tf_rpc=ApplyResourceChange tf_attribute_path=auto_scale tf_resource_type=fabric_spark_custom_pool @caller=github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwschemadata/value_semantic_equality.go:91 timestamp="2025-02-06T14:39:08.679+0100"
2025-02-06T14:39:08.680+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: Value switched to prior value due to semantic equality logic: tf_attribute_path=dynamic_executor_allocation.enabled tf_req_id=ID tf_provider_addr=registry.terraform.io/microsoft/fabric tf_resource_type=fabric_spark_custom_pool tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwschemadata/value_semantic_equality.go:91 @module=sdk.framework timestamp="2025-02-06T14:39:08.679+0100"
2025-02-06T14:39:08.680+0100 [DEBUG] provider.terraform-provider-fabric_v0.1.0-beta.8.exe: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_attribute_path=type tf_provider_addr=registry.terraform.io/microsoft/fabric tf_rpc=ApplyResourceChange tf_req_id=ID tf_resource_type=fabric_spark_custom_pool @caller=github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwschemadata/value_semantic_equality.go:91 timestamp="2025-02-06T14:39:08.679+0100"
2025-02-06T14:39:08.711+0100 [DEBUG] State storage *statemgr.Filesystem declined to persist a state snapshot
2025-02-06T14:39:08.711+0100 [ERROR] vertex "fabric_spark_custom_pool.example" error: Provider produced inconsistent result after apply
2025-02-06T14:39:08.712+0100 [ERROR] vertex "fabric_spark_custom_pool.example" error: Provider produced inconsistent result after apply
2025-02-06T14:39:08.734+0100 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2025-02-06T14:39:08.763+0100 [INFO]  provider: plugin process exited: plugin=.terraform/providers/registry.terraform.io/microsoft/fabric/0.1.0-beta.8/windows_amd64/terraform-provider-fabric_v0.1.0-beta.8.exe id=ID
```


### 📷 Screenshots

_No response_

### 📈 Expected behavior

an updated fabric_spark_custom_pool

### 🌌 Environment (Provider Version)

0.1.0-beta.8

### 🌌 Environment (Terraform Version)

1.10.5

### 🌌 Environment (OS)

Windows

### 📎 Additional context

_No response_

### 🔰 Code of Conduct

- [x] I agree to follow this project's Code of Conduct.

--- a/.changes/unreleased/fixed-20250206-214937.yaml
+++ b/.changes/unreleased/fixed-20250206-214937.yaml
@@ -1,0 +1,5 @@
+kind: fixed
+body: Inconsistent result for dynamic_executor_allocation (min_executors/max_executors) when enabled is false.
+time: 2025-02-06T21:49:37.8248633+01:00
+custom:
+  Issue: "239"

--- a/docs/resources/spark_custom_pool.md
+++ b/docs/resources/spark_custom_pool.md
@@ -78,6 +78,9 @@ Required:
 Required:
 
 - `enabled` (Boolean) The status of the dynamic executor allocation. Accepted values: `false` - Disabled, `true` - Enabled.
+
+Optional:
+
 - `max_executors` (Number) The maximum executors.
 - `min_executors` (Number) The minimum executors.
 

--- a/internal/services/spark/base_test.go
+++ b/internal/services/spark/base_test.go
@@ -19,28 +19,6 @@ const (
 	sparkEnvironmentLibrariesTFName = spark.SparkEnvironmentLibrariesTFName
 )
 
-func getSparkCustomPoolResourceAttr(t *testing.T, workspaceID, name string) map[string]any {
-	t.Helper()
-
-	return map[string]any{
-		"workspace_id": workspaceID,
-		"name":         name,
-		"type":         "Workspace",
-		"node_family":  "MemoryOptimized",
-		"node_size":    "Small",
-		"auto_scale": map[string]any{
-			"enabled":        true,
-			"min_node_count": 1,
-			"max_node_count": 3,
-		},
-		"dynamic_executor_allocation": map[string]any{
-			"enabled":       true,
-			"min_executors": 1,
-			"max_executors": 2,
-		},
-	}
-}
-
 func environmentResource(t *testing.T, workspaceID string) (resourceHCL, resourceFQN string) {
 	t.Helper()
 

--- a/internal/services/spark/data_spark_custom_pool_test.go
+++ b/internal/services/spark/data_spark_custom_pool_test.go
@@ -32,7 +32,23 @@ func TestAcc_SparkCustomPoolDataSource(t *testing.T) {
 				workspaceResourceHCL,
 				at.CompileConfig(
 					testResourceSparkCustomPoolHeader,
-					getSparkCustomPoolResourceAttr(t, testhelp.RefByFQN(workspaceResourceFQN, "id"), entityName),
+					map[string]any{
+						"workspace_id": testhelp.RefByFQN(workspaceResourceFQN, "id"),
+						"name":         entityName,
+						"type":         "Workspace",
+						"node_family":  "MemoryOptimized",
+						"node_size":    "Small",
+						"auto_scale": map[string]any{
+							"enabled":        true,
+							"min_node_count": 1,
+							"max_node_count": 3,
+						},
+						"dynamic_executor_allocation": map[string]any{
+							"enabled":       true,
+							"min_executors": 1,
+							"max_executors": 2,
+						},
+					},
 				),
 				at.CompileConfig(
 					testDataSourceSparkCustomPoolHeader,

--- a/internal/services/spark/resource_spark_custom_pool.go
+++ b/internal/services/spark/resource_spark_custom_pool.go
@@ -11,6 +11,7 @@ import (
 
 	"github.com/hashicorp/terraform-plugin-framework-timeouts/resource/timeouts"
 	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
 	"github.com/hashicorp/terraform-plugin-framework/diag"
 	"github.com/hashicorp/terraform-plugin-framework/path"
 	"github.com/hashicorp/terraform-plugin-framework/resource"
@@ -18,10 +19,12 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
 	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
 	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
 	"github.com/hashicorp/terraform-plugin-log/tflog"
 	fabcore "github.com/microsoft/fabric-sdk-go/fabric/core"
 	fabspark "github.com/microsoft/fabric-sdk-go/fabric/spark"
 	supertypes "github.com/orange-cloudavenue/terraform-plugin-framework-supertypes"
+	superint32validator "github.com/orange-cloudavenue/terraform-plugin-framework-validators/int32validator"
 
 	"github.com/microsoft/terraform-provider-fabric/internal/common"
 	"github.com/microsoft/terraform-provider-fabric/internal/framework/customtypes"
@@ -130,11 +133,33 @@ func (r *resourceSparkCustomPool) Schema(ctx context.Context, _ resource.SchemaR
 					},
 					"min_executors": schema.Int32Attribute{
 						MarkdownDescription: "The minimum executors.",
-						Required:            true,
+						Computed:            true,
+						Optional:            true,
+						Validators: []validator.Int32{
+							superint32validator.NullIfAttributeIsOneOf(
+								path.MatchRoot("dynamic_executor_allocation").AtName("enabled"),
+								[]attr.Value{types.BoolValue(false)},
+							),
+							superint32validator.RequireIfAttributeIsOneOf(
+								path.MatchRoot("dynamic_executor_allocation").AtName("enabled"),
+								[]attr.Value{types.BoolValue(true)},
+							),
+						},
 					},
 					"max_executors": schema.Int32Attribute{
 						MarkdownDescription: "The maximum executors.",
-						Required:            true,
+						Computed:            true,
+						Optional:            true,
+						Validators: []validator.Int32{
+							superint32validator.NullIfAttributeIsOneOf(
+								path.MatchRoot("dynamic_executor_allocation").AtName("enabled"),
+								[]attr.Value{types.BoolValue(false)},
+							),
+							superint32validator.RequireIfAttributeIsOneOf(
+								path.MatchRoot("dynamic_executor_allocation").AtName("enabled"),
+								[]attr.Value{types.BoolValue(true)},
+							),
+						},
 					},
 				},
 			},

--- a/internal/services/spark/resource_spark_custom_pool_test.go
+++ b/internal/services/spark/resource_spark_custom_pool_test.go
@@ -22,15 +22,8 @@ func TestAcc_SparkCustomPoolResource_CRUD(t *testing.T) {
 	capacityID := capacity["id"].(string)
 
 	workspaceResourceHCL, workspaceResourceFQN := testhelp.TestAccWorkspaceResource(t, capacityID)
-	testHelperSparkCustomPoolResource := getSparkCustomPoolResourceAttr(t, testhelp.RefByFQN(workspaceResourceFQN, "id"), "test")
-
 	entityCreateName := testhelp.RandomName()
-	testCaseCreate := testhelp.CopyMap(testHelperSparkCustomPoolResource)
-	testCaseCreate["name"] = entityCreateName
-
 	entityUpdateName := testhelp.RandomName()
-	testCaseUpdate := testhelp.CopyMap(testHelperSparkCustomPoolResource)
-	testCaseUpdate["name"] = entityUpdateName
 
 	resource.Test(t, testhelp.NewTestAccCase(t, &testResourceSparkCustomPoolFQN, nil, []resource.TestStep{
 		// Create and Read
@@ -40,10 +33,32 @@ func TestAcc_SparkCustomPoolResource_CRUD(t *testing.T) {
 				workspaceResourceHCL,
 				at.CompileConfig(
 					testResourceSparkCustomPoolHeader,
-					testCaseCreate,
+					map[string]any{
+						"workspace_id": testhelp.RefByFQN(workspaceResourceFQN, "id"),
+						"name":         entityCreateName,
+						"type":         "Workspace",
+						"node_family":  "MemoryOptimized",
+						"node_size":    "Small",
+						"auto_scale": map[string]any{
+							"enabled":        true,
+							"min_node_count": 1,
+							"max_node_count": 3,
+						},
+						"dynamic_executor_allocation": map[string]any{
+							"enabled":       true,
+							"min_executors": 1,
+							"max_executors": 2,
+						},
+					},
 				)),
 			Check: resource.ComposeAggregateTestCheckFunc(
 				resource.TestCheckResourceAttr(testResourceSparkCustomPoolFQN, "name", entityCreateName),
+				resource.TestCheckResourceAttr(testResourceSparkCustomPoolFQN, "auto_scale.enabled", "true"),
+				resource.TestCheckResourceAttr(testResourceSparkCustomPoolFQN, "auto_scale.min_node_count", "1"),
+				resource.TestCheckResourceAttr(testResourceSparkCustomPoolFQN, "auto_scale.max_node_count", "3"),
+				resource.TestCheckResourceAttr(testResourceSparkCustomPoolFQN, "dynamic_executor_allocation.enabled", "true"),
+				resource.TestCheckResourceAttr(testResourceSparkCustomPoolFQN, "dynamic_executor_allocation.min_executors", "1"),
+				resource.TestCheckResourceAttr(testResourceSparkCustomPoolFQN, "dynamic_executor_allocation.max_executors", "2"),
 			),
 		},
 		// Update and Read
@@ -53,10 +68,30 @@ func TestAcc_SparkCustomPoolResource_CRUD(t *testing.T) {
 				workspaceResourceHCL,
 				at.CompileConfig(
 					testResourceSparkCustomPoolHeader,
-					testCaseUpdate,
+					map[string]any{
+						"workspace_id": testhelp.RefByFQN(workspaceResourceFQN, "id"),
+						"name":         entityUpdateName,
+						"type":         "Workspace",
+						"node_family":  "MemoryOptimized",
+						"node_size":    "Small",
+						"auto_scale": map[string]any{
+							"enabled":        false,
+							"min_node_count": 1,
+							"max_node_count": 3,
+						},
+						"dynamic_executor_allocation": map[string]any{
+							"enabled": false,
+						},
+					},
 				)),
 			Check: resource.ComposeAggregateTestCheckFunc(
 				resource.TestCheckResourceAttr(testResourceSparkCustomPoolFQN, "name", entityUpdateName),
+				resource.TestCheckResourceAttr(testResourceSparkCustomPoolFQN, "auto_scale.enabled", "false"),
+				resource.TestCheckResourceAttr(testResourceSparkCustomPoolFQN, "auto_scale.min_node_count", "1"),
+				resource.TestCheckResourceAttr(testResourceSparkCustomPoolFQN, "auto_scale.max_node_count", "3"),
+				resource.TestCheckResourceAttr(testResourceSparkCustomPoolFQN, "dynamic_executor_allocation.enabled", "false"),
+				resource.TestCheckNoResourceAttr(testResourceSparkCustomPoolFQN, "dynamic_executor_allocation.min_executors"),
+				resource.TestCheckNoResourceAttr(testResourceSparkCustomPoolFQN, "dynamic_executor_allocation.max_executors"),
 			),
 		},
 	},

